### PR TITLE
Add Request Monitoring Feature to CoAP Connection

### PR DESCRIPTION
**Pull request by jkralik**
This pull request introduces a new feature to the CoAP connection module - Request Monitoring. The `WithRequestMonitor` function has been implemented to enable request monitoring for the connection. This function is called for each CoAP message received from the peer before it is processed.

Details of the Feature:
- **Functionality:** `WithRequestMonitor` allows developers to implement custom request monitoring logic for incoming CoAP messages.
- **Error Handling:** If the function returns an error, the connection is closed, providing a mechanism to handle and respond to issues in the monitoring process.
- **Message Dropping:** If the function returns true, the incoming message is dropped, allowing for selective handling or filtering of messages based on monitoring criteria.

This new feature enhances the flexibility and control developers have over the CoAP connection behavior. Please review the changes and provide feedback.

### CODE REVIEW

**Comment by jkralik** [comment]
@jeffwelder-ellenbytech I have updated the code as we wanted to merge it. Additionally, I will add some tests there.

**Comment by JeffWelder** [comment]
@jkralik Great, I pulled the latest commit into my project to test, and it works as expected. :)

File: tcp/client_test.go [hunk 1]
```diff
@@ -667,3 +668,167 @@ func TestClientKeepAliveMonitor(t *testing.T) {
...
+	require.NoError(t, err)
+	require.Equal(t, codes.Content.String(), got.Code().String())
+
+	// New request but with DELETE code to trigger EOF error from request monitor
+	deleteReq, err := cc.NewDeleteRequest(ctx, "/test")
+	require.NoError(t, err)
+	deleteReq.SetMessageID(2)
+	_, err = cc.Do(deleteReq)
+	require.Error(t, err)
+	<-reqMonitorErr
```
Comment (Danielius1922) [inline tcp/client_test.go:755]: Maybe

```go
select {
case <-reqMonitorErr:
case <-ctx.Done():
  t.FailNow()
}
```

to guarantee that the test won't get stuck

Comment (jkralik) [reply tcp/client_test.go]: Fixed

**Review by Danielius1922 — APPROVED** [review]

**Pull request merged**